File: nlp.js
/**
 * NLP 模块 - 轻量级中英文分词
 * 功能：分词、词频统计、共现关系提取
 */

// 常见中文词汇字典 (常用 2-4 字词)
const COMMON_CHINESE_WORDS = new Set([
  // 常见名词
  '苹果', '水果', '手机', '电脑', '网络', '软件', '硬件', '系统', '数据', '信息',
  '技术', '科学', '学习', '教育', '学校', '学生', '老师', '工作', '公司', '产品',
  '服务', '用户', '客户', '市场', '价格', '质量', '时间', '空间', '世界', '中国',
  '美国', '日本', '欧洲', '亚洲', '国际', '政治', '经济', '文化', '历史', '社会',
  '生命', '水', '空气', '光', '热', '温度', '能量', '物质', '分子', '原子',
  // 常见动词
  '学习', '工作', '生活', '使用', '开发', '设计', '创造', '管理', '组织', '计划',
  '开始', '结束', '进行', '完成', '发展', '变化', '增长', '减少', '提高', '降低',
  // 常见形容词
  '重要', '简单', '复杂', '困难', '容易', '快速', '慢速', '高效', '低效', '现代',
  '传统', '新', '旧', '大', '小', '长', '短', '高', '低', '好', '坏',
  // 常见虚词 (保留用于上下文)
  '什么', '怎么', '如何', '为什么', '因为', '所以', '但是', '如果', '虽然', '只是'
]);

// 停用词列表
const STOP_WORDS = new Set([
  // 中文停用词
  '的', '是', '在', '了', '和', '与', '或', '有', '这', '那', '个', '一', '不', '也',
  '都', '就', '而', '及', '以', '对', '可', '能', '会', '被', '于', '从', '到', '把',
  '将', '为', '但', '却', '又', '如', '因', '所', '并', '其', '之', '来', '去', '上',
  '下', '中', '大', '小', '多', '少', '最', '更', '很', '太', '过', '要', '该', '我们',
  '你们', '他们', '她们', '它们', '这个', '那个', '可以', '没有', '这样', '那样',
  '自己', '已经', '因为', '所以', '但是', '而且', '或者', '如果', '虽然', '只是',
  '就是', '还是', '应该', '需要', '可能', '关于', '什么', '怎么', '如何', '为什么',
  // 英文停用词
  'the', 'a', 'an', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has',
  'had', 'do', 'does', 'did', 'will', 'would', 'could', 'should', 'may', 'might',
  'must', 'shall', 'can', 'need', 'dare', 'ought', 'used', 'to', 'of', 'in', 'for',
  'on', 'with', 'at', 'by', 'from', 'as', 'into', 'through', 'during', 'before',
  'after', 'above', 'below', 'between', 'under', 'again', 'further', 'then', 'once',
  'and', 'but', 'or', 'nor', 'so', 'yet', 'both', 'either', 'neither', 'not', 'only',
  'own', 'same', 'than', 'too', 'very', 'just', 'also', 'now', 'here', 'there', 'when',
  'where', 'why', 'how', 'all', 'each', 'every', 'few', 'more', 'most', 'other', 'some',
  'such', 'no', 'any', 'what', 'which', 'who', 'whom', 'this', 'that', 'these', 'those',
  'it', 'its', 'i', 'me', 'my', 'we', 'our', 'you', 'your', 'he', 'she', 'him', 'her', 'his'
]);

/**
 * 智能分词 - 混合中英文
 * 策略：优先匹配常见词汇，然后使用 n-gram
 */
function tokenize(text) {
  if (!text || typeof text !== 'string') return [];

  const tokens = [];
  const seen = new Set();

  // 1. 尝试匹配常见中文词汇 (2-4字)
  let remaining = text;
  const wordMatches = [];

  // 先找出所有可能匹配的位置
  COMMON_CHINESE_WORDS.forEach(word => {
    let pos = 0;
    while (true) {
      const idx = remaining.indexOf(word, pos);
      if (idx === -1) break;
      wordMatches.push({ word, start: idx, end: idx + word.length });
      pos = idx + 1;
    }
  });

  // 按起始位置排序
  wordMatches.sort((a, b) => a.start - b.start);

  // 提取匹配的词 (避免重叠)
  const matchedRanges = [];
  wordMatches.forEach(match => {
    const overlaps = matchedRanges.some(r =>
      (match.start >= r.start && match.start < r.end) ||
      (match.end > r.start && match.end <= r.end)
    );
    if (!overlaps) {
      matchedRanges.push(match);
      if (!seen.has(match.word)) {
        seen.add(match.word);
        tokens.push(match.word);
      }
    }
  });

  // 2. 提取英文单词 (2+字母)
  const englishRegex = /[a-zA-Z]{2,}/g;
  const englishWords = text.match(englishRegex) || [];
  englishWords.forEach(w => {
    const lower = w.toLowerCase();
    if (!STOP_WORDS.has(lower) && !seen.has(lower)) {
      seen.add(lower);
      tokens.push(lower);
    }
  });

  // 3. 提取数字
  const numberRegex = /\d+/g;
  const numbers = text.match(numberRegex) || [];
  numbers.forEach(n => {
    if (n.length >= 2 && !seen.has(n)) {
      seen.add(n);
      tokens.push(n);
    }
  });

  // 4. 对未覆盖的中文字符，使用 2-gram
  let chineseOnly = text;
  matchedRanges.forEach(r => {
    chineseOnly = chineseOnly.replace(r.word, ' '.repeat(r.word.length));
  });

  // 移除已处理的英文字母和数字
  chineseOnly = chineseOnly.replace(/[a-zA-Z0-9]/g, ' ');

  // 提取剩余的中文 2-gram
  const chineseNgram = chineseOnly.match(/[\u4e00-\u9fff]{2}/g) || [];
  chineseNgram.forEach(gram => {
    if (!STOP_WORDS.has(gram) && !seen.has(gram)) {
      seen.add(gram);
      tokens.push(gram);
    }
  });

  // 5. 额外：尝试提取 3-gram 关键词
  const chinese3gram = text.match(/[\u4e00-\u9fff]{3}/g) || [];
  chinese3gram.forEach(gram => {
    if (!STOP_WORDS.has(gram) && !seen.has(gram) && !COMMON_CHINESE_WORDS.has(gram)) {
      // 只保留看起来像词的 (非随机组合)
      if (gram.includes('的') || gram.includes('了') || gram.includes('是')) {
        // 包含虚词的跳过
      } else {
        seen.add(gram);
        tokens.push(gram);
      }
    }
  });

  return tokens;
}

/**
 * 过滤停用词
 */
function filterStopWords(tokens) {
  return tokens.filter(token => {
    if (token.length < 2) return false;
    if (STOP_WORDS.has(token.toLowerCase())) return false;
    return true;
  });
}

/**
 * 统计词频
 */
function countWordFrequency(texts, topN = 20) {
  const frequency = {};

  texts.forEach(text => {
    const tokens = tokenize(text);
    tokens.forEach(token => {
      frequency[token] = (frequency[token] || 0) + 1;
    });
  });

  const sorted = Object.entries(frequency)
    .sort((a, b) => b[1] - a[1])
    .slice(0, topN);

  return Object.fromEntries(sorted);
}

/**
 * 计算共现关系
 */
function extractCoOccurrences(targetConcept, texts, minFrequency = 1) {
  const coOccurrences = {};

  texts.forEach(text => {
    const tokens = tokenize(text);
    const hasTarget = tokens.some(t => t.includes(targetConcept) || targetConcept.includes(t));

    if (hasTarget) {
      tokens.forEach(token => {
        if (!token.includes(targetConcept) && !targetConcept.includes(token)) {
          coOccurrences[token] = (coOccurrences[token] || 0) + 1;
        }
      });
    }
  });

  return Object.fromEntries(
    Object.entries(coOccurrences).filter(([_, count]) => count >= minFrequency)
  );
}

/**
 * 提取关键词
 */
function extractKeywords(texts, targetConcept, topN = 15) {
  const wordFreq = countWordFrequency(texts, topN * 2);
  const coOccurrences = extractCoOccurrences(targetConcept, texts);

  const keywords = {};
  const allWords = new Set([...Object.keys(wordFreq), ...Object.keys(coOccurrences)]);

  allWords.forEach(word => {
    const freqScore = wordFreq[word] || 0;
    const coocScore = coOccurrences[word] || 0;
    keywords[word] = freqScore * 0.4 + coocScore * 0.6;
  });

  const sorted = Object.entries(keywords)
    .sort((a, b) => b[1] - a[1])
    .slice(0, topN);

  return sorted.map(([word, score]) => ({ word, score: Math.round(score * 100) / 100 }));
}

/**
 * 提取上下文
 */
function extractContexts(texts, targetConcept, contextLength = 50) {
  const contexts = [];

  texts.forEach(text => {
    const index = text.indexOf(targetConcept);
    if (index !== -1) {
      const start = Math.max(0, index - contextLength);
      const end = Math.min(text.length, index + targetConcept.length + contextLength);
      let context = text.substring(start, end);

      if (start > 0) context = '...' + context;
      if (end < text.length) context = context + '...';

      contexts.push(context);
    }
  });

  return contexts.slice(0, 5);
}

module.exports = {
  tokenize,
  filterStopWords,
  countWordFrequency,
  extractCoOccurrences,
  extractKeywords,
  extractContexts,
  STOP_WORDS
};

File: cli/src/nlp.rs
// NLP Module - Tokenization, Stop Words, Word Frequency

use lazy_static::lazy_static;
use regex::Regex;
use std::collections::HashSet;

lazy_static! {
    // Common Chinese words dictionary
    static ref COMMON_CHINESE: HashSet<&'static str> = {
        let mut s = HashSet::new();
        // Common nouns
        s.insert("苹果"); s.insert("水果"); s.insert("手机"); s.insert("电脑");
        s.insert("网络"); s.insert("软件"); s.insert("硬件"); s.insert("系统");
        s.insert("数据"); s.insert("信息"); s.insert("技术"); s.insert("科学");
        s.insert("学习"); s.insert("教育"); s.insert("学校"); s.insert("学生");
        s.insert("老师"); s.insert("工作"); s.insert("公司"); s.insert("产品");
        s.insert("服务"); s.insert("用户"); s.insert("客户"); s.insert("市场");
        s.insert("价格"); s.insert("质量"); s.insert("时间"); s.insert("空间");
        s.insert("世界"); s.insert("中国"); s.insert("美国"); s.insert("日本");
        s.insert("欧洲"); s.insert("亚洲"); s.insert("国际"); s.insert("政治");
        s.insert("经济"); s.insert("文化"); s.insert("历史"); s.insert("社会");
        s.insert("生命"); s.insert("水"); s.insert("空气"); s.insert("光");
        s.insert("热"); s.insert("温度"); s.insert("能量"); s.insert("物质");
        s.insert("分子"); s.insert("原子");
        // Common verbs
        s.insert("学习"); s.insert("工作"); s.insert("生活"); s.insert("使用");
        s.insert("开发"); s.insert("设计"); s.insert("创造"); s.insert("管理");
        s.insert("组织"); s.insert("计划"); s.insert("开始"); s.insert("结束");
        s.insert("进行"); s.insert("完成"); s.insert("发展"); s.insert("变化");
        s.insert("增长"); s.insert("减少"); s.insert("提高"); s.insert("降低");
        // Common adjectives
        s.insert("重要"); s.insert("简单"); s.insert("复杂"); s.insert("困难");
        s.insert("容易"); s.insert("快速"); s.insert("慢速"); s.insert("高效");
        s.insert("低效"); s.insert("现代"); s.insert("传统");
        // Question words
        s.insert("什么"); s.insert("怎么"); s.insert("如何"); s.insert("为什么");
        s
    };

    // Stop words (Chinese + English)
    static ref STOP_WORDS: HashSet<&'static str> = {
        let mut s = HashSet::new();
        // Chinese stop words
        s.insert("的"); s.insert("是"); s.insert("在"); s.insert("了"); s.insert("和");
        s.insert("与"); s.insert("或"); s.insert("有"); s.insert("这"); s.insert("那");
        s.insert("个"); s.insert("一"); s.insert("不"); s.insert("也"); s.insert("都");
        s.insert("就"); s.insert("而"); s.insert("及"); s.insert("以"); s.insert("对");
        s.insert("可"); s.insert("能"); s.insert("会"); s.insert("被"); s.insert("于");
        s.insert("从"); s.insert("到"); s.insert("把"); s.insert("将"); s.insert("为");
        s.insert("但"); s.insert("却"); s.insert("又"); s.insert("如"); s.insert("因");
        s.insert("所"); s.insert("并"); s.insert("其"); s.insert("之"); s.insert("来");
        s.insert("去"); s.insert("上"); s.insert("下"); s.insert("中"); s.insert("大");
        s.insert("小"); s.insert("多"); s.insert("少"); s.insert("最"); s.insert("更");
        s.insert("很"); s.insert("太"); s.insert("过"); s.insert("要"); s.insert("该");
        s.insert("我们"); s.insert("你们"); s.insert("他们"); s.insert("她们");
        s.insert("它们"); s.insert("这个"); s.insert("那个"); s.insert("可以");
        s.insert("没有"); s.insert("这样"); s.insert("那样"); s.insert("自己");
        s.insert("已经"); s.insert("因为"); s.insert("所以"); s.insert("但是");
        s.insert("而且"); s.insert("或者"); s.insert("如果"); s.insert("虽然");
        s.insert("只是"); s.insert("就是"); s.insert("还是"); s.insert("应该");
        s.insert("需要"); s.insert("可能"); s.insert("关于");
        // English stop words
        s.insert("the"); s.insert("a"); s.insert("an"); s.insert("is"); s.insert("are");
        s.insert("was"); s.insert("were"); s.insert("be"); s.insert("been"); s.insert("being");
        s.insert("have"); s.insert("has"); s.insert("had"); s.insert("do"); s.insert("does");
        s.insert("did"); s.insert("will"); s.insert("would"); s.insert("could"); s.insert("should");
        s.insert("may"); s.insert("might"); s.insert("must"); s.insert("shall"); s.insert("can");
        s.insert("need"); s.insert("dare"); s.insert("ought"); s.insert("used"); s.insert("to");
        s.insert("of"); s.insert("in"); s.insert("for"); s.insert("on"); s.insert("with");
        s.insert("at"); s.insert("by"); s.insert("from"); s.insert("as"); s.insert("into");
        s.insert("through"); s.insert("during"); s.insert("before"); s.insert("after");
        s.insert("above"); s.insert("below"); s.insert("between"); s.insert("under");
        s.insert("again"); s.insert("further"); s.insert("then"); s.insert("once");
        s.insert("and"); s.insert("but"); s.insert("or"); s.insert("nor"); s.insert("so");
        s.insert("yet"); s.insert("both"); s.insert("either"); s.insert("neither");
        s.insert("not"); s.insert("only"); s.insert("own"); s.insert("same"); s.insert("than");
        s.insert("too"); s.insert("very"); s.insert("just"); s.insert("also"); s.insert("now");
        s.insert("here"); s.insert("there"); s.insert("when"); s.insert("where"); s.insert("why");
        s.insert("how"); s.insert("all"); s.insert("each"); s.insert("every"); s.insert("few");
        s.insert("more"); s.insert("most"); s.insert("other"); s.insert("some"); s.insert("such");
        s.insert("no"); s.insert("any"); s.insert("what"); s.insert("which"); s.insert("who");
        s.insert("whom"); s.insert("this"); s.insert("that"); s.insert("these"); s.insert("those");
        s.insert("it"); s.insert("its"); s.insert("i"); s.insert("me"); s.insert("my");
        s.insert("we"); s.insert("our"); s.insert("you"); s.insert("your"); s.insert("he");
        s.insert("she"); s.insert("him"); s.insert("her"); s.insert("his");
        s
    };
}

/// Tokenize text (Chinese + English mixed)
pub fn tokenize(text: &str) -> Vec<String> {
    if text.is_empty() {
        return vec![];
    }

    let mut tokens = Vec::new();
    let mut seen = std::collections::HashSet::new();

    // 1. Extract English words (2+ letters)
    if let Ok(re) = Regex::new(r"[a-zA-Z]{2,}") {
        for cap in re.find_iter(text) {
            let word = cap.as_str().to_lowercase();
            if !STOP_WORDS.contains(word.as_str()) && !seen.contains(&word) && word.len() >= 2 {
                seen.insert(word.clone());
                tokens.push(word);
            }
        }
    }

    // 2. Extract Chinese 2-gram
    if let Ok(re) = Regex::new(r"[\u4e00-\u9fff]{2}") {
        for cap in re.find_iter(text) {
            let word = cap.as_str();
            if !STOP_WORDS.contains(word) && !seen.contains(word) {
                seen.insert(word.to_string());
                tokens.push(word.to_string());
            }
        }
    }

    // 3. Extract Chinese 3-gram (keywords)
    if let Ok(re) = Regex::new(r"[\u4e00-\u9fff]{3}") {
        for cap in re.find_iter(text) {
            let word = cap.as_str();
            // Skip if contains function words
            if word.contains('的') || word.contains('了') || word.contains('是') {
                continue;
            }
            if !STOP_WORDS.contains(word) && !seen.contains(word) {
                seen.insert(word.to_string());
                tokens.push(word.to_string());
            }
        }
    }

    // 4. Extract Chinese 4-gram (common words)
    if let Ok(re) = Regex::new(r"[\u4e00-\u9fff]{4}") {
        for cap in re.find_iter(text) {
            let word = cap.as_str();
            if COMMON_CHINESE.contains(word) && !seen.contains(word) {
                seen.insert(word.to_string());
                tokens.push(word.to_string());
            }
        }
    }

    tokens
}

/// Filter stop words from tokens
pub fn filter_stop_words(tokens: &[String]) -> Vec<String> {
    tokens
        .iter()
        .filter(|t| t.len() >= 2 && !STOP_WORDS.contains(t.as_str()))
        .cloned()
        .collect()
}

/// Count word frequency
pub fn count_frequency(texts: &[String], top_n: usize) -> Vec<(String, u32)> {
    let mut freq = std::collections::HashMap::new();

    for text in texts {
        let tokens = tokenize(text);
        for token in tokens {
            *freq.entry(token).or_insert(0) += 1;
        }
    }

    let mut items: Vec<_> = freq.into_iter().collect();
    items.sort_by(|a, b| b.1.cmp(&a.1));
    items.truncate(top_n);
    items
}

/// Extract co-occurrences with target concept
pub fn extract_cooccurrences(target: &str, texts: &[String], min_freq: u32) -> Vec<(String, u32)> {
    let mut cooc = std::collections::HashMap::new();

    for text in texts {
        let tokens = tokenize(text);
        let has_target = tokens.iter().any(|t| t.contains(target) || target.contains(t));

        if has_target {
            for token in &tokens {
                if !token.contains(target) && !target.contains(token) {
                    *cooc.entry(token.clone()).or_insert(0) += 1;
                }
            }
        }
    }

    let mut items: Vec<_> = cooc.into_iter().filter(|(_, v)| *v >= min_freq).collect();
    items.sort_by(|a, b| b.1.cmp(&a.1));
    items
}

/// Extract keywords by combining frequency and co-occurrence
pub fn extract_keywords(texts: &[String], target: &str, top_n: usize) -> Vec<(String, f64)> {
    let word_freq = count_frequency(texts, top_n * 2);
    let cooc = extract_cooccurrences(target, texts, 1);

    let mut keywords = std::collections::HashMap::new();
    let all_words: std::collections::HashSet<_> = word_freq.iter()
        .map(|(w, _)| w)
        .chain(cooc.iter().map(|(w, _)| w))
        .collect();

    for word in all_words {
        let freq_score = word_freq.iter().find(|(w, _)| w == word).map(|(_, v)| *v).unwrap_or(0) as f64;
        let cooc_score = cooc.iter().find(|(w, _)| w == word).map(|(_, v)| *v).unwrap_or(0) as f64;
        keywords.insert(word.clone(), freq_score * 0.4 + cooc_score * 0.6);
    }

    let mut items: Vec<_> = keywords.into_iter().collect();
    items.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
    items.truncate(top_n);
    items
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tokenize() {
        let text = "水是地球上最常见的物质之一。水的化学式是H2O。";
        let tokens = tokenize(text);
        println!("{:?}", tokens);
        assert!(tokens.len() > 0);
    }

    #[test]
    fn test_filter_stop_words() {
        let tokens = vec!["水".to_string(), "的".to_string(), "是".to_string(), "重要".to_string()];
        let filtered = filter_stop_words(&tokens);
        assert!(filtered.contains(&"水".to_string()));
        assert!(!filtered.contains(&"的".to_string()));
    }
}